packages/lib/test/computedTree/computedTree.test.ts

```diff
@@ -0,0 +1,518 @@
+    expect(resolvePath(r, ["model"])).toEqual({ resolved: true, value: r.model })
+  })
+
+  test("computed tree supports findParent", () => {
+    const r = new R({})
+
+    const isR = (node: unknown) => node instanceof R
+
+    expect(findParent(r.array, isR)).toBe(r)
+    expect(findParent(r.plainObject, isR)).toBe(r)
+    expect(findParent(r.model, isR)).toBe(r)
+  })
+
+  test("computed tree supports findParentPath", () => {
+    const r = new R({})
+
+    const isR = (node: unknown) => node instanceof R
+
+    expect(findParentPath(r.array, isR)).toEqual({ parent: r, path: ["array"] })
+    expect(findParentPath(r.plainObject, isR)).toEqual({ parent: r, path: ["plainObject"] })
+    expect(findParentPath(r.model, isR)).toEqual({ parent: r, path: ["model"] })
+  })
+
+  test("computed tree supports findChildren", () => {
+    const r = new R({})
+
+    // cached when observed
+    autoDispose(
+      reaction(
+        () => [r.array, r.plainObject, r.model],
+        () => {}
+      )
+    )
+
+    expect(findChildren(r, (node) => isArray(node)).has(r.array)).toBeTruthy()
+    expect(findChildren(r, (node) => isPlainObject(node)).has(r.plainObject)).toBeTruthy()
+    expect(findChildren(r, (node) => node instanceof M).has(r.model)).toBeTruthy()
+  })
+
+  test.each([false, true])("computed tree supports getChildrenObjects (deep=%j)", (deep) => {
+    const r = new R({})
+
+    // no children objects because the computed trees are not observed at this point
+    expect(getChildrenObjects(r, { deep }).size).toBe(0)
+
+    // cached when observed
+    autoDispose(
+      reaction(
+        () => [r.array, r.plainObject, r.model],
+        () => {}
+      )
+    )
+
+    const children = getChildrenObjects(r, { deep })
+    expect(children.size).toBe(3)
+    expect(children.has(r.array)).toBeTruthy()
+    expect(children.has(r.plainObject)).toBeTruthy()
+    expect(children.has(r.model)).toBeTruthy()
+  })
+
+  test.each([WalkTreeMode.ChildrenFirst, WalkTreeMode.ParentFirst])(
+    "computed tree supports walkTree (mode=%s)",
+    (mode) => {
+      const r = new R({})
+
+      // does not traverse computed trees because they are not obesrved at this point
+      expect(walkTree(r, (node) => (isArray(node) ? node : undefined), mode)).toBeUndefined()
+      expect(walkTree(r, (node) => (isPlainObject(node) ? node : undefined), mode)).toBeUndefined()
+      expect(walkTree(r, (node) => (node instanceof M ? node : undefined), mode)).toBeUndefined()
+
+      // cached when observed
+      autoDispose(
+        reaction(
+          () => [r.array, r.plainObject, r.model],
+          () => {}
+        )
+      )
+
+      expect(walkTree(r, (node) => (isArray(node) ? node : undefined), mode)).toBe(r.array)
+      expect(walkTree(r, (node) => (isPlainObject(node) ? node : undefined), mode)).toBe(
+        r.plainObject
+      )
+      expect(walkTree(r, (node) => (node instanceof M ? node : undefined), mode)).toBe(r.model)
```

Comment:
hmm, I'm not sure if it is desirable for computed trees to appear in walkTree at all...

Reply:
perhaps isComputedTreeNode could be used before deciding if the callback should be called?

Reply:
Hm, if `walkTree` didn't call the callback for computed tree nodes, then the `onAttachedToRootStore` hook wouldn't be called for those nodes either because [root store attachment/detachment uses `walkTree` internally](https://github.com/xaviergonz/mobx-keystone/blob/9c11ff21d64512ca3bc3ceb5a953eab2cc4a51f3/packages/lib/src/rootStore/attachDetach.ts#L22-L43). We _could_ introduce an option like `includeComputedTreeNodes: boolean` and exclude computed tree nodes by default for backwards compatibility. In the root store attachment/detachment code, it would need to be set to `true` though which (I think) introduces some coupling between the root store feature and the computed tree feature.

Also, my view on computed trees is that they are essentially equivalent to regular trees except they are immutable and, thus, action middlewares are useless because there are no action calls in computed trees. With that in mind, I think they should be treated no different than regular trees as much as possible, so e.g. tree traversal functions should behave the same for computed and regular trees.

But I'm more than happy to be convinced otherwise. :slightly_smiling_face:

Reply:
Oh, and if `walkTree` filtered out computed tree nodes, then it would be impossible to call it on a computed tree node as the subtree root, too, so it would be entirely useless in computed trees.

Reply:
> We could introduce an option like includeComputedTreeNodes: boolean and exclude computed tree nodes by default for backwards compatibility

I think that would make sense, or maybe even an enum: 'nonComputed' | 'computed' | 'all'